server/src/main/java/com/microsoft/java/bs/core/internal/services/LifecycleService.java

```diff
@@ -119,61 +134,117 @@ public void exit() {
-   * <ul>
-   *   <li>Gradle Java home is not set.</li>
-   *   <li>A valid JDK can be found to launch Gradle.</li>
-   * </ul>
-   *
-   * <p>The JDK installation path string will be set to {@link Preferences#gradleJavaHome}.
+   * Finds a suitable version of JDK to use for compilation of Java programs.
+   * Returns {@code null} if no suitable JDK can be found.
    */
-  private void updateGradleJavaHomeIfNecessary(URI rootUri) {
+  private File getSuitableJdk(URI rootUri) {
+
     Preferences preferences = preferenceManager.getPreferences();
-    if (preferences.getJdks() == null || preferences.getJdks().isEmpty()) {
-      return;
+    String filePath = preferences.getGradleJavaHome();
+    File gradleJavaHome;
+    if (StringUtils.isNotEmpty(filePath)) {
+      gradleJavaHome = new File(filePath);
+    } else {
+      gradleJavaHome = connector.getGradleJavaHome(rootUri);
     }
 
-    if (StringUtils.isBlank(preferences.getGradleJavaHome())) {
-      String gradleVersion = "";
-      GradleBuildKind buildKind = Utils.getEffectiveBuildKind(new File(rootUri), preferences);
-      Map<String, String> map = TelemetryUtils.getMetadataMap("buildKind", buildKind.name());
-      LOGGER.log(Level.INFO, "Use build kind: " + buildKind.name(), map);
-      if (buildKind == GradleBuildKind.SPECIFIED_VERSION) {
-        gradleVersion = preferences.getGradleVersion();
-      } else {
-        gradleVersion = connector.getGradleVersion(rootUri);
-      }
+    GradleBuildKind buildKind = Utils.getEffectiveBuildKind(new File(rootUri), preferences);
+    Map<String, String> map = TelemetryUtils.getMetadataMap("buildKind", buildKind.name());
+    LOGGER.log(Level.INFO, "Use build kind: " + buildKind.name(), map);
+
+    String gradleVersion;
+    if (buildKind == GradleBuildKind.SPECIFIED_VERSION) {
+      gradleVersion = preferences.getGradleVersion();
+    } else {
+      gradleVersion = connector.getGradleVersion(rootUri);
+    }
+
+    if (StringUtils.isEmpty(gradleVersion)) {
+      return null;
+    }
+
+    map = TelemetryUtils.getMetadataMap("gradleVersion", gradleVersion);
+    LOGGER.log(Level.INFO, "Gradle version: " + gradleVersion, map);
+
+    String latestCompatibleVersion = Utils.getLatestCompatibleJavaVersion(gradleVersion);
+    String oldestCompatibleVersion = Utils.getOldestCompatibleJavaVersion();
+
+    if (StringUtils.isNotBlank(latestCompatibleVersion)) {
 
-      if (StringUtils.isNotBlank(gradleVersion)) {
-        map = TelemetryUtils.getMetadataMap("gradleVersion", gradleVersion);
-        LOGGER.log(Level.INFO, "Gradle version: " + gradleVersion, map);
-        String highestJavaVersion = Utils.getHighestCompatibleJavaVersion(gradleVersion);
-        File jdkInstallation = getJdkToLaunchDaemon(preferences.getJdks(), highestJavaVersion);
-        if (jdkInstallation != null) {
-          preferences.setGradleJavaHome(jdkInstallation.getAbsolutePath());
+      // Use GradleJavaHome if compatible
+      if (gradleJavaHome != null) {
```

Comment:
Will `gradleJavaHome` null here? Given we have logic at line 152

Reply:
Yes, it will be null if the `GradleApiConnector#getGradleJavaHome` gets a `BuildException`. In this case we will get a null value.

Reply:
Ok, my next question is that will `GradleApiConnector#getGradleJavaHome` gurantee that the found java home is compatible?

Reply:
No. It is only going to return the `GradleJavaHome`, looking for it in the priority order listed in [gradle docs](https://docs.gradle.org/current/userguide/build_environment.html#priority_for_configurations). It only queries for a `GradleJavaHome` and does check for compatibility. It can even return invalid filepath if the user has wrong configurations for the project (eg. invalid jdk path in gradle.properties or JAVA_HOME).